api/src/modules/renown/kyc-service.ts

```diff
@@ -0,0 +1,201 @@
+  ceramicUrl?: string;
+  definition: RuntimeCompositeDefinition
+}
+
+export class KYCService {
+
+  protected client: ComposeClient;
+
+  protected eventSource: EventSource | null = null;
+
+  protected entries: any[] = [];
+
+  protected codec: Codec<any, any, any>;
+
+  constructor(options: KYCServiceOptions) {
+    const ceramic = options.ceramicUrl || "http://localhost:7007";
+    this.client = new ComposeClient({
+      ceramic,
+      definition: options.definition as RuntimeCompositeDefinition,
+    });
+
+    this.eventSource = new EventSource(`${ceramic}/api/v0/feed/aggregation/documents`)
+    this.codec = JsonAsString.pipe(AggregationDocument)
+  }
+
+  #fetchEntriesFromCeramic = async (first: number, skip: number) => {
+    const date = new Date().toISOString()
+    const entries = await this.client.executeQuery<{
+      verifiableCredentialEIP712Index: {
+        edges: { node: CeramicPowerhouseVerifiableCredential }[];
+      };
+    }>(fetchEntries, {
+      first,
+      skip,
+      input: {
+        and: [
+          {
+            or: [
+              {
+                where: {
+                  revocationDate: {
+                    greaterThan: date,
+                  },
+                },
+              },
+              {
+                where: {
+                  revocationDate: {
+                    isNull: true,
+                  },
+                },
+              },
+            ],
+          },
+        ],
+      },
+    });
+
+    return entries as any[];
+  }
+
+  #fetchEntryFromCeramic = async (issuerId: string, subjectId: string) => {
+    const date = new Date().toISOString()
+    const result = await this.client.executeQuery<{
+      verifiableCredentialEIP712Index: {
+        edges: { node: CeramicPowerhouseVerifiableCredential }[];
+      };
+    }>(fetchEntries, {
+      input: {
+        and: [
+          {
+            where: {
+              issuerId: {
+                equalTo: issuerId,
+              },
+              subjectId: {
+                equalTo: subjectId,
+              },
+            },
+          },
+          {
+            or: [
+              {
+                where: {
+                  revocationDate: {
+                    greaterThan: date,
+                  },
+                },
+              },
+              {
+                where: {
+                  revocationDate: {
+                    isNull: true,
+                  },
+                },
+              },
+            ],
+          },
+        ],
+      },
+    });
+
+    if (result.errors) {
+      return null;
+    }
+
+    const { data } = result;
+    if (!data || data.verifiableCredentialEIP712Index.edges.length === 0) {
+      return null;
+    }
+
+    return data.verifiableCredentialEIP712Index.edges[0]!.node;
+  }
+
+  #updateEntries = async () => {
+    const first = 100;
+    let skip = 0;
+    let entries: any[] = [];
+    let foundEntries = 0;
+    const fetchEntriesLoop = async () => {
```

Comment:
I would use `while` loop instead. Each recursive call is added to the stack, and only when they are all finished are they removed. For the near future there will not be enough credentials to cause a stack overflow but I don't see an advantage in using a recursive call anyway.

I also noticed that recursive calls are harder to debug, when something happens in `processNextJob` on the queues, the callstack is a bunch of `processNextJob` calls which makes it hard to understand the actual caller.

Reply:
I did it this way, because I got a linting error when I used await inside of a while loop.